astroplan/constraints.py

```diff
@@ -0,0 +1,565 @@
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
+"""
+Specify and constraints to determine which targets are observable for
+an observer.
+"""
+
+from __future__ import (absolute_import, division, print_function,
+                        unicode_literals)
+
+# Standard library
+from abc import ABCMeta, abstractmethod
+import datetime
+
+# Third-party
+from astropy.time import Time
+import astropy.units as u
+from astropy.coordinates import get_sun, Angle
+from astropy.coordinates.angle_utilities import angular_separation
+from astropy.extern.six import string_types
+import numpy as np
+
+# Package
+from .moon import get_moon, moon_illumination
+
+DEFAULT_TIME_RESOLUTION = 0.5*u.hour
+
+__all__ = ["AltitudeConstraint", "AirmassConstraint", "AtNightConstraint",
+           "is_observable", "is_always_observable", "time_grid_from_range",
+           "SunSeparationConstraint", "MoonSeparationConstraint",
+           "MoonIlluminationConstraint", "LocalTimeConstraint"]
+
+@u.quantity_input(time_resolution=u.hour)
+def time_grid_from_range(time_range, time_resolution=DEFAULT_TIME_RESOLUTION):
+    """
+    Get linearly-spaced sequence of times.
+
+    Parameters
+    ----------
+    time_range : `~astropy.time.Time`
+        Lower and upper bounds on time sequence. If a scalar time is input,
+        make the input time the lower limit on the time sequence, and the
+        upper bound one attosecond later than the input time.
+
+    time_resolution : `~astropy.units.quantity` (optional)
+        Time-grid spacing
+
+    Returns
+    -------
+    times : `~astropy.time.Time`
+        Linearly-spaced sequence of times
+    """
+    if time_range.isscalar:
+        return Time([time_range])
+    else:
+        return Time(np.arange(time_range[0].jd, time_range[1].jd,
+                              time_resolution.to(u.day).value), format='jd')
+
+class Constraint(object):
+    """
+    Abstract class for objects defining observational constraints.
+    """
+    __metaclass__ = ABCMeta
+
+    def __call__(self, time_range, observer, targets,
+                 time_resolution=DEFAULT_TIME_RESOLUTION):
+
+        if not isinstance(time_range, Time):
+            time_range = Time(time_range)
+        cons = self._compute_constraint(time_range, observer, targets,
+                                        time_resolution=time_resolution)
+        return cons
+
+    def _compute_constraint(self, time_range, observer, targets):
```

Comment:
I think this method should get the ``@abstractmethod`` decorator.  That way, if a user tries to do ``Constraint()``, they see an error right away telling them this is an abstract base class.